[XLA:CPU][oneDNN] Adding support for Quantization

mahmoud-abuzaina opened the pull request:
- Depends on https://github.com/openxla/xla/pull/19066 and https://github.com/openxla/xla/pull/19067

- This PR adds support for quantization in XLA according to the design explained in the quantization [RFC](https://docs.google.com/document/d/1c9Yfd5OnPiORHku1ELbUDdojz7kKmStNRMjEFjF-pQ0/edit?tab=t.0)

- Note about the change in cpu_compiler.cc file: the simplification pass has been moved from running after onednn rewriter to run before it. This simplifies the patterns for fusing the dequantize sub-graph as it gets rid of redundant Copy instructions introduced be earlier passes.

mahmoud-abuzaina committed:
Adding support for Quantization using oneDNN

sdasgup3 commented on xla/tests/hlo_test_base.h:
```diff
@@ -148,8 +148,10 @@ class HloTestBase : public ::testing::Test {
   // HLO matches the given FileCheck pattern.
-  void MatchOptimizedHlo(absl::string_view hlo, absl::string_view pattern,
-                         bool print_operand_shape = false);
+  void MatchOptimizedHlo(
```
Can you please update the comment to reflect the extra argument added?

mahmoud-abuzaina replied:
Done.

sdasgup3 commented on xla/service/cpu/onednn_contraction_rewriter.cc:
```diff
@@ -470,6 +477,242 @@ bool OneDnnContractionRewriter::ShouldRewriteConv(
+  if (quant_scale && dequant_scale) {
+    auto quant_scale_value = GetScalarConstantValue(quant_scale);
+    auto dequant_scale_value = GetScalarConstantValue(dequant_scale);
+    if (quant_scale_value && dequant_scale_value) {
```
Should we consder the case where one of [quant_scale_value, dequant_scale_value] is null and other not?

same for the zp check below.

mahmoud-abuzaina replied:
yes, I did some refactoring to properly check for scale/zp matching.

mahmoud-abuzaina committed:
Resolved merge conflicts

mahmoud-abuzaina committed:
Addressed review comments

mahmoud-abuzaina commented:
@sdasgup3 Thanks for the comments. I addressed them. Also, I created a separate [PR](https://github.com/openxla/xla/pull/18441) to fix a bug that was exposed by this PR due to changes in cpu_compiler.cc

penpornk commented on xla/service/cpu/onednn_util.h:
```diff
@@ -23,14 +23,40 @@ limitations under the License.
 namespace xla {
 namespace cpu {
 
+using dnnl::engine;
+using dnnl::memory;
```
Per [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Aliases), we don't use using aliases or typedefs in header files just for convenience. (Only when they are intended to be part of the API.) 

Please use the full type instead. (Can still use using aliases in .cc file or private scopes.)

mahmoud-abuzaina replied:
Fixed.

penpornk commented on xla/service/cpu/onednn_util.h:
```diff
@@ -59,7 +90,7 @@ dnnl::stream MakeOneDnnStream(
     dnnl::threadpool_interop::threadpool_iface* thread_pool);
 
-typedef BackendConfig::BackendConfigOneofCase BackendConfigOneofCase;
+typedef BackendConfig::BackendConfigOneofCase BackendConfigOneofCase; 
```
Please remove `typedef` and use the full type in header file.

mahmoud-abuzaina replied:
Fixed.

penpornk reviewed (changes requested):
Sorry for the delay! This PR can be refactored to smaller PRs. I'm still thinking about the best way to do it. In the meanwhile, we can 
1. Get https://github.com/openxla/xla/pull/18441 in.
2. Put the cpu_compiler.cc change in another PR. We are reordering the pass so it would be good to isolate this change and see if something fails.
3. Put the changes in onednn_memory_util.cc in its own PR (and a test case that catches the scalar case).

mahmoud-abuzaina committed:
Merged with main branch

mahmoud-abuzaina committed:
Addressed review comments

mahmoud-abuzaina commented on xla/service/cpu/onednn_contraction_rewriter.cc:
```diff
@@ -833,6 +1222,30 @@ class OneDnnContractionRewriteVisitor : public DfsHloRewriteVisitor {
     return absl::OkStatus();
   }
 
+  // This is a heuristic check to check if Multiply op is possibly part of
```
Fixed.

mahmoud-abuzaina commented on xla/service/cpu/tests/onednn_convolution_test.cc:
Rebased.

mahmoud-abuzaina commented:
> Sorry for the delay! This PR can be refactored to smaller PRs. I'm still thinking about the best way to do it. In the meanwhile, we can
> 
> 1. Get [[XLA:CPU][oneDNN] Fixing a bug in algebraic simplifier #18441](https://github.com/openxla/xla/pull/18441) in.
> 2. Put the cpu_compiler.cc change in another PR. We are reordering the pass so it would be good to isolate this change and see if something fails.
> 3. Put the changes in onednn_memory_util.cc in its own PR (and a test case that catches the scalar case).

Thank you for reviewing the PR. I have addressed the review comments and created two additional PRs as suggested, #19066 and #19067.

mahmoud-abuzaina committed:
Resolved merge conflicts

penpornk commented:
> Thank you for reviewing the PR. I have addressed the review comments and created two additional PRs as suggested, #19066 and #19067.

Thank you! Let's get #19066 and #19067 in first before we get back to this PR.

mahmoud-abuzaina committed:
Resolved merge conflicts

dimitar-asenov commented:
https://github.com/openxla/xla/pull/19067 Was rolled back. We should wait for it to be merged before proceeding with the current PR.

dimitar-asenov commented:
This seems to be still be blocked on the other PR that was rolled back.
